src/Control/Monad/Bayes/Sampler/Lazy.hs

```diff
@@ -67,5 +71,44 @@ sampler m = newStdGen *> (runSampler m . randomTree <$> getStdGen)
 independent :: Monad m => m a -> m [a]
 independent = sequence . repeat
 
+-- | 'weightedsamples' runs a probability measure and gets out a stream of (result,weight) pairs
 weightedsamples :: Weighted Sampler a -> IO [(a, Log Double)]
 weightedsamples = sampler . independent . weighted
+
+wiener :: Sampler (Double -> Double)
+wiener = Sampler $ \(Tree _ gs) ->
+  unsafePerformIO $ do
+    ref <- newIORef M.empty
+    modifyIORef' ref (M.insert 0 0)
+    return \x -> unsafePerformIO do
+      table <- readIORef ref
+      case M.lookup x table of
+        Just y -> return y
+        Nothing -> do
+          let lower = do
+                l <- findMaxLower x (M.keys table)
+                v <- M.lookup l table
+                return (l, v)
+          let upper = do
+                u <- find (> x) (M.keys table)
+                v <- M.lookup u table
+                return (u, v)
+          let m = bridge lower x upper
+          let y = runSampler m (gs !! (1 + M.size table))
+          modifyIORef' ref (M.insert x y)
+          return y
+  where
+    bridge :: Maybe (Double, Double) -> Double -> Maybe (Double, Double) -> Sampler Double
+    -- not needed since the table is always initialized with (0, 0)
+    -- bridge Nothing y Nothing = if y==0 then return 0 else normal 0 (sqrt y)
+    bridge (Just (x, x')) y Nothing = normal x' (sqrt (y - x))
+    bridge Nothing y (Just (z, z')) = normal z' (sqrt (z - y))
+    bridge (Just (x, x')) y (Just (z, z')) = normal (x' + ((y - x) * (z' - x') / (z - x))) (sqrt ((z - y) * (y - x) / (z - x)))
+    bridge _ _ _ = error "undefined behavior for bridge"
+
+    findMaxLower :: Double -> [Double] -> Maybe Double
```

Comment:
Alternative proposition of the implementation - may be easier to read:
```
findMaxLower :: Double -> [Double] -> Maybe Double
findMaxLower d = foldl' (upd d) Nothing
  where
    upd limit acc x = if x >= limit then acc
                      else maximum (acc, Just x)
```

Reply:
That seems smart